sql-plugin/src/main/scala/com/nvidia/spark/rapids/SortUtils.scala

```diff
@@ -46,4 +56,211 @@ object SortUtils {
     (order.isAscending && order.nullOrdering == NullsFirst) ||
       (!order.isAscending && order.nullOrdering == NullsLast)
   }
+
+  @scala.annotation.tailrec
+  def extractReference(exp: Expression): Option[GpuBoundReference] = exp match {
+    case r: GpuBoundReference => Some(r)
+    case a: Alias => extractReference(a.child)
+    case _ => None
+  }
+
+  def getOrder(order: SortOrder, index: Int): Table.OrderByArg =
+    if (order.isAscending) {
+      Table.asc(index, order.nullOrdering == NullsFirst)
+    } else {
+      Table.desc(index, order.nullOrdering == NullsLast)
+    }
+}
+
+class GpuSorter(
+    sortOrder: Seq[SortOrder],
+    inputSchema: Array[Attribute]) extends Arm with Serializable {
+
+  def this(sortOrder: Seq[SortOrder], inputSchema: Seq[Attribute]) =
+    this(sortOrder, inputSchema.toArray)
+
+  private[this] val boundSortOrder = GpuBindReferences.bindReferences(sortOrder, inputSchema.toSeq)
+
+  private[this] val numInputColumns = inputSchema.length
+
+  def cpuOrdering: Seq[SortOrder] = cpuOrderingInternal.toSeq
+
+  private[this] lazy val (needsComputation, cudfOrdering, cpuOrderingInternal) = {
+    val needsComputation = mutable.ArrayBuffer[SortOrder]()
+    val cpuOrdering = mutable.ArrayBuffer[SortOrder]()
+    val cudfOrdering = mutable.ArrayBuffer[Table.OrderByArg]()
+    var newColumnIndex = numInputColumns
+    // Remove duplicates in the ordering itself because
+    // There is no need to do it twice
+    boundSortOrder.distinct.foreach { so =>
+      SortUtils.extractReference(so.child) match {
+        case Some(ref) =>
+          cudfOrdering += SortUtils.getOrder(so, ref.ordinal)
+          // It is a bound GPU reference so we have to translate it to the CPU
+          cpuOrdering += ShimLoader.getSparkShims.sortOrder(
+            BoundReference(ref.ordinal, ref.dataType, ref.nullable),
+            so.direction, so.nullOrdering)
+        case None =>
+          val index = newColumnIndex
+          newColumnIndex += 1
+          cudfOrdering += SortUtils.getOrder(so, index)
+          needsComputation += so
+          // We already did the computation so instead of trying to translate
+          // the computation back to the CPU too, just use the existing columns.
+          cpuOrdering += ShimLoader.getSparkShims.sortOrder(
+            BoundReference(index, so.dataType, so.nullable),
+            so.direction, so.nullOrdering)
+      }
+    }
+    (needsComputation.toArray, cudfOrdering.toArray, cpuOrdering.toArray)
+  }
+
+  lazy val computationSchema: Seq[Attribute] =
+    needsComputation.map { so =>
+      AttributeReference(s"SORT_TMP", so.dataType, so.nullable)()
```

Comment:
I'm not following the use of `SORT_TMP`.  A block comment at the top of the class describing at a high level what the algorithm is and how extra columns are used would help.

Reply:
Would still like to see a block comment on the class describing the high-level algorithm being performed (i.e.: mentioning the fact that there are extra columns, how they're computed and used, maybe walk through an example, etc.)